AlsoInThis side panel follow-up

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

https://user-images.githubusercontent.com/6356129/141417838-65d424dd-5263-43b6-9848-07d0c8e73450.mp4

- Title can now be 2 lines @jtamiace 
- Less `position:absolute`, refactored layout a bit for simpler approach to accommodate the 2-line request
- CSS cleaned up quite a bit
- Use `useContentNodeProgress` various tools to build and use a shared store (:heart_eyes:) 
- General clean up, improved comments, etc

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

I don't see an issue to fix, but this is follow up on #8651 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

The above is what I intended to do here. There was discussion around routing that should be handled in a separate follow-up PR. 

Copy-pasted from #8651 - all should still work:

It ended up being more complicated than I had anticipated but it should be good to go :)

Mainly be sure that you can get the

General

Progress indicator on content where there is progress. Just a star if completed
Scrollbar appears in largely populated folders and all items are visible when scrolled down
The content you're viewing doesn't get listed in the side bar
Content which has a duration will show the time under the title
A channel with lots of folders, if you view a content a level or two deep, then you should be guaranteed to see the link on the bottom of the sidebar linking to the folder.
Lesson context

The contents listed should be from within the same lesson, not from siblings of that content in it's other locations (in channels in the library).
Links should remain in Lesson context - so you should always see the "view lesson resources" icon as you click through side panel "next in lesson" resources
Mobile

The progress / completed icon appears underneath the title (and optionally the duration if present)

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
refactored styles for AlsoInThis per feedback

nucleogenesis committed:
use useContentNodeProgress to make things far, far, far simpler

nucleogenesis committed:
no console.logs

rtibbles commented on kolibri/plugins/learn/assets/src/views/AlsoInThis.vue:
```diff
@@ -24,31 +24,34 @@
+          </div>
+          <div class="progress">
+            <KIcon
+              v-if="withProgress(content).progress === 1"
```
The progress bar already contains this conditional logic: https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/plugins/learn/assets/src/views/ProgressBar.vue#L11

And accesses progress via the contentNodeMap from the useContentNodeProgress composable - so no need to mix it into the content node first.

Can also remove this addition from the previous PR: https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/plugins/learn/assets/src/views/ProgressBar.vue#L60

nucleogenesis replied:
Thanks for the reminder - once I got the bits hooked up elsewhere and they worked I was too pleased to remember I had to make proper use of ProgressBar

rtibbles commented on kolibri/plugins/learn/assets/src/views/LearnImmersiveLayout.vue:
```diff
@@ -275,38 +273,22 @@
+       * is logged in
        */
       fetchLessonSiblings() {
         LessonResource.fetchModel({ id: this.lessonId }).then(lesson => {
```
Can just call `this.fetchContentNodeProgress({ lesson_id: this.lessonId });` here in parallel to the lesson fetch to avoid chaining (with the isUserLoggedIn check).

rtibbles commented on kolibri/plugins/learn/assets/src/views/LearnImmersiveLayout.vue:
```diff
@@ -315,27 +297,30 @@
+            const removeSelfFilter = node => node.id !== this.content.id;
+            if (this.$store.getters.isUserLoggedIn) {
+              this.mapSidePanelContentProgress(contentNodes).then(
+                () => (this.viewResourcesContents = contentNodes.filter(removeSelfFilter))
```
I would suggest treating the contentNodeProgress load as a lazy load (this is how we treat it in other places in the codebase), and not block loading of the main content for it.

If you just rely on the built in ProgressBar functionality as I suggest above, it will reactively update when the progress data has been loaded.

rtibbles commented on kolibri/plugins/learn/assets/src/views/LearnImmersiveLayout.vue:
```diff
@@ -154,7 +151,8 @@
       const { canDownload } = useCoreLearn();
-      return { canDownload };
+      const { contentNodeProgressMap, fetchContentNodeProgress } = useContentNodeProgress();
+      return { canDownload, contentNodeProgressMap, fetchContentNodeProgress };
```
Not seeing where `contentNodeProgressMap` is used in the component.

rtibbles commented on kolibri/plugins/learn/assets/src/views/LearnImmersiveLayout.vue:
```diff
@@ -315,27 +297,30 @@
-
         ContentNodeResource.fetchTree({
           id: this.content.parent,
           params: {
```
If you're only interested in the immediate children, can also pass `depth: 1` in the params, to avoid recursing to the grand children of the parent.

rtibbles reviewed (dismissed):
I think the progress fetching and tracking can be made lazy and some of the previously added complexity removed.

I think for the sake of simplicity and not waiting on chaining, just using direct params filtering for the content node progress fetch (rather than waiting for nodes to return then fetching ids) will make for a snappier user experience.

nucleogenesis committed:
Remove errant addition of a .progress expectation in ProgressBar

nucleogenesis committed:
LearnImmersiveLayout - lazier progress mapping, improved fetching, clean

- Fetch and map progress via useContentNodeProgress lazily
- Use lessonId when present as param for fetching progress
- Add `depth: 1` param to fetching of siblings (not lesson context) to save the backend some work

nucleogenesis committed:
86 withProgress() in AlsoInThis, check progress for mastered differently

- pass content node to progress bar and let it handle everything from there when the content isn't mastered

rtibbles commented on kolibri/plugins/learn/assets/src/views/AlsoInThis.vue:
```diff
@@ -24,31 +24,34 @@
+            />
+          </div>
+          <div class="progress">
+            <KIcon
```
This `KIcon` is also inside the `ProgressBar` component - is there a styling or layout reason to not to defer to it here? If so, would be good to have a code comment to explain.

rtibbles commented on kolibri/plugins/learn/assets/src/views/LearnImmersiveLayout.vue:
```diff
@@ -275,81 +273,70 @@
+            const getParams = {
+              ids: this.viewResourcesContents.map(node => node.id),
+            };
+            this.fetchContentNodeProgress(getParams);
```
Can avoid the delayed load here still by using `fetchContentNodeTreeProgress`:
```
this.fetchContentNodeTreeProgress({
          id: this.content.parent,
          params: {
            include_coach_content:
              this.$store.getters.isAdmin ||
              this.$store.getters.isCoach ||
              this.$store.getters.isSuperuser,
            depth: 1,
          },
        })
```

Can then be invoked in parallel to the node fetching.

rtibbles reviewed (commented):
A couple of comments but nothing blocking code wise. Have not reviewed functionality!

jtamiace reviewed (approved):
Was able to get all the scenarios listed working. Thank you for doing the follow up!

jtamiace commented:
Actually... I think I found a bug. If you click on a resource from the Bookmarks tab, and that resource was bookmarked from a Channel, the 'Also in this' panel shows a random lesson that the bookmarked resource is not a part of.

I realize this happens in a specific sequence:
1. Go to a resource in a lesson
2. Exit the lesson
3. Go to the bookmarks tab
4. Go to a bookmarked resource
5. Open in the 'Also in this' panel

It should be the lesson list. But if you: (1) Exit the bookmarked resouce renderer, (2) refresh the page, (3) click the bookmarked resource again, (4) open the 'Also in this' panel

It should return to normal

rtibbles commented:
I am going to file this as a follow up issue, as this was pre-existing to this PR, I think.
